proposals/slottable-request.md

```diff
@@ -0,0 +1,86 @@
+# Slottable Request Protocol
+aka "Render Props for Web Components"
+
+**Author**: Kevin Schaaf (@kevinpschaaf, Google)
+
+**Status**: Draft
+
+**Last Updated**: 2023-08-28
+
+
+# Background
+
+There are often use cases that require a component to render UI based on data, but where the specific rendering of the data should be controllable by the user. Use cases of this include:
+
+*   A data table or tree component, which manages the rendering of cells or tree nodes, but where the given look and feel of the items is customizable by the user.
+*   A virtual list component, which optimizes rendering a list of data by only rendering a subset of user-templated items based on which array instances are visible on the screen.
+*   Components that fetch data themselves, but provide the ability for the user to customize the rendering of the data.
+
+This proposal describes an interoperable protocol for components to request that their owning context render slotted content into themselves, using data provided via the protocol.
+
+# Goals
+
+*   The ability for users of a component to render requested content using a templating system of their choice, using data supplied by the component.
+*   The ability for requested content to be logically composed by the requesting component into the component's shadow root at a position of its choice.
+*   The ability for requested content to be styled in the same "user" scope the component in question was created in, and using styling mechanisms of the user's choice.
+*   The ability to implement the protocol inside framework-specific wrappers or helpers that expose a user API that looks similar or identical to framework-specific patterns (i.e. "render props" for React).
+
+
+# Design / Proposal
+
+
+## Protocol Definition
+
+Concretely, the protocol involves dispatching a well-known DOM event to request slotted content (aka "[slottables](https://dom.spec.whatwg.org/#light-tree-slotables)", per the DOM spec) be rendered with the provided `data` argument and assigned to the given `slotName`:
+
+```ts
+export class SlotableRequestEvent<Name extends string, T = unknown> extends Event {
+  readonly data: T | typeof remove;
+  readonly name: Name;
+  readonly slotName: string;
+  constructor(name: Name, data: unknown, key?: string) {
+    super('slottable-request', {bubbles: false, composed: false});
+    this.name = name;
+    this.data = data;
+    this.slotName = key !== undefined ? `${name}.${key}` : name;
+  }
+}
+
+export const remove = Symbol('slottable-request-remove');
+```
+
+When a component wants to render customizable content, it (1) fires a `slottable-request` event for a given slot name, and (2) renders a `<slot>` element into its shadow root corresponding to the requested slot name.
```

Comment:
What happens if the component is defined before the parent component has been defined? It will miss these events.

One solution here is to have an imperative API, such as a function that you can call, to cause a render to occur.

Reply:
Could you explain who would call the imperative API, and how they would know call it?

Slotting is really a form of dependency injection, and it does get tricky when the owning scope isn't available to provide dependencies. We've dealt with this in the context protocol at an implementation level by having an "unfulfilled context request" manager at the root re-dispatch context requests from consumers to providers [see discussion](https://github.com/webcomponents-cg/community-protocols/issues/25#issuecomment-1105862016). We haven't standardized that into the protocol, but we implement it in Lit to pretty good effect, and I could see something like that forming a solution.

Reply:
It's unclear to me how the event-listener is intended to provide the content. Inline examples would help clarify this for me.